pkg/apis/volumesnapshot/v1beta1/types.go

```diff
@@ -0,0 +1,354 @@
+// TODO(xiangqian): Add a webhook to ensure that VolumeSnapshotSource members
+// will not be updated once specified.
+type VolumeSnapshotSource struct {
+	// persistentVolumeClaimName specifies the name of the PersistentVolumeClaim
+	// object in the same namespace as the VolumeSnapshot object where the
+	// snapshot should be dynamically taken from.
+	// +optional
+	PersistentVolumeClaimName *string `json:"persistentVolumeClaimName,omitempty" protobuf:"bytes,1,opt,name=persistentVolumeClaimName"`
+
+	// volumeSnapshotContentName specifies the name of a pre-existing VolumeSnapshotContent
+	// object a user asks to statically bind the VolumeSnapshot object to.
+	// +optional
+	VolumeSnapshotContentName *string `json:"volumeSnapshotContentName,omitempty" protobuf:"bytes,2,opt,name=volumeSnapshotContentName"`
+}
+
+// VolumeSnapshotStatus is the status of the VolumeSnapshot
+type VolumeSnapshotStatus struct {
+	// NOTE: All fields in VolumeSnapshotStatus are informational for user references.
+	// Controllers MUST NOT rely on any fields programmatically.
+
+	// boundVolumeSnapshotContentName represents the name of the VolumeSnapshotContent
+	// object to which the VolumeSnapshot object is bound.
+	// If not specified, it indicates that the VolumeSnapshot object has not been
+	// successfully bound to a VolumeSnapshotContent object yet.
+	// +optional
+	BoundVolumeSnapshotContentName *string `json:"boundVolumeSnapshotContentName,omitempty" protobuf:"bytes,1,opt,name=boundVolumeSnapshotContentName"`
+
+	// creationTime, if not nil, represents the timestamp when the point-in-time
+	// snapshot was successfully cut on the underlying storage system.
+	// In dynamic snapshot creation case, it will be filled in upon snapshot creation.
+	// For a pre-existing snapshot, it will be filled in once the VolumeSnapshot object has
+	// been successfully bound to a VolumeSnapshotContent object and the underlying
+	// storage system has the information available.
+	// If not specified, it indicates that the creation time of the snapshot is unknown.
+	// +optional
+	CreationTime *metav1.Time `json:"creationTime,omitempty" protobuf:"bytes,2,opt,name=creationTime"`
+
+	// readyToUse indicates if a snapshot is ready to be used to restore a volume.
+	// In dynamic snapshot creation case, readyToUse will be set to true after underlying storage
+	// system has successfully finished all out-of-bound procedures to make a snapshot ready to
+	// be used to restore a volume.
+	// For a pre-existing snapshot, readyToUse will be set to the value returned
+	// from CSI "ListSnapshots" gRPC call if the matching CSI driver exists and supports.
+	// Otherwise, if there exists no CSI driver or the driver does not support "ListSnapshots",
+	// this field will be set to "True".
+	// If not specified, it indicates that the readiness of a snapshot is unknown.
+	// +optional
+	ReadyToUse *bool `json:"readyToUse,omitempty" protobuf:"varint,3,opt,name=readyToUse"`
```

Comment:
Is there any other use case other than restore? Should the field be called "ReadyForRestore"?

Reply:
For this ReadyToUse field, is that possible that we add some logic in webhook to verify the bi-directional pointers and content ReadyToUse status each time it is updated to set to "true". If that is possible, could we avoid user/controller to explicitly verify the pointers each time they use snapshot?

Reply:
Users may decide not to run the webhook. I think it is better if controllers are resilient.

Reply:
ok, then we cannot just rely on webhook to verify. Maybe it still worth to add this logic in webhook to have some extra protection. But we still put important note in the comment to remind users to verify pointers.

Reply:
webhooks are great for preventing operational errors, i.e., kubectl apply -f wrongfile. for subresource status update, in this case, status in VolumeSnapshot,
it will be updated by common controller, and binding will be verified before any updates. I think that should be enough?

Reply:
the verification is to prevent some malicious attacker to set up their own controller to update this field, I think

Reply:
Still outstanding question on naming of this field. Would it make more sense to call it "ReadyForRestore" to make it more explicit what use case it is ready for? "Use" is ambiguous

Reply:
discussed on this, we will keep the name align with CSI returned "read_to_use"